### PR TITLE
implement UV_EXTERN_VISIBILITY_DEFAULT for GCC and Clang

### PR DESCRIPTION
**Pull request by wrl**
Previously, the `visibility` variable in common.gypi only controlled whether GCC or Clang builds would build with -fvisibility=hidden, but unfortunately this effect was negated by a define in uv.h setting UV_EXTERN to mark symbols with `default` visibility. This change sets -fvisibility=hidden as default and adds a define for setting UV_EXTERN to `default` visibility, which also enables building with `hidden` visibility, for example as a static library.

### CODE REVIEW

**Comment by wrl** [comment]
i'm in the CLA now. not sure what subsystem this falls under, but if that and the commit line length are huge issues for you, let me know and i'll amend the commit.

**Comment by indutny** [comment]
Thank you! I would place it under `gyp` subsystem, and yeah we care about commit line length. Perhaps you may omit `GCC and Clang` in it?

**Comment by indutny** [comment]
@wrl ping ;)

**Comment by wrl** [comment]
@indutny how's this?

File: common.gypi [hunk 1]
```diff
@@ -154,7 +154,7 @@
             'cflags': [ '-pthread' ],
             'ldflags': [ '-pthread' ],
           }],
-          [ 'visibility=="hidden" and (clang==1 or gcc_version >= 40)', {
+          [ '(clang==1 or gcc_version >= 40)', {
```
Comment (indutny) [inline common.gypi:157]: Parens are not required.

File: include/uv.h [hunk 2]
```diff
@@ -39,7 +39,7 @@
     /* Building static library. */
 #   define UV_EXTERN /* nothing */
 # endif
-#elif __GNUC__ >= 4
+#elif __GNUC__ >= 4 && defined(UV_EXTERN_VISIBILITY_DEFAULT)
```
Comment (bnoordhuis) [inline include/uv.h:42]: What about the autotools and Makefile.mingw builds?

Comment (indutny) [inline include/uv.h:42]: @wrl that is a good point too ^

Comment (wrl) [inline include/uv.h:42]: @bnoordhuis yeah, very good point. currently they don't do anything about visibility (no `-fvisibility=hidden` anywhere).
i could either flip this conditional so `default` is...well, the default, and switch it to `UV_EXTERN_VISIBILITY_HIDDEN`, or define `UV_EXTERN_VISIBILITY_DEFAULT` in the autotools and mingw builds. which would you prefer?

win32 has its own separate issues, however (uses shared vs static builds to determine symbol visibility), and i was directed by @indutny to not worry about it initially.

Comment (indutny) [inline include/uv.h:42]: @wrl it'd be good to have some sort of switch for `UV_EXTERN_VISIBILITY` in autotools too.

About mingw... I'm not sure about how it actually works, but I think that it is still creating dlls for shared libraries. Which means that the config we already have should work in that case too.

**Comment by indutny** [comment]
@wrl ping

**Comment by saghul** [comment]
No follow-up, closing.

**Pull request closed without merging**